File: pbwebwcl/src/app/gestione-economie/services/gestione-economie.service.ts
import { DOCUMENT } from '@angular/common';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Inject, Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { ConfigService } from 'src/app/core/services/config.service';
import { BandoLinea } from '../commons/dto/bando-linea';
import { BeneficiarioQuoteDTO } from '../commons/dto/beneficiario-quote-dto';
import { DatiGeneraliDTO } from '../commons/dto/dati-generali-dto';
import { EconomiaDTO } from '../commons/dto/economia-dto';
import { EsitoOperazioneEconomia } from '../commons/dto/esito-operazione-economia';
import { ModificaDTO } from '../commons/dto/modifica-dto';
import { ParametriRicercaEconomieDTO } from '../commons/dto/parametri-ricerca-economia-dto';
import { ProgettoDTO } from '../commons/dto/progetto-dto';

@Injectable()
export class GestioneEconomieService {

    constructor(
        private configService: ConfigService,
        private http: HttpClient
    ) { }

    getEconomie(idU: number, request: ParametriRicercaEconomieDTO): Observable<Array<EconomiaDTO>> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/geteconomie';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.post<Array<EconomiaDTO>>(url, request, { params: params });
    }

    deleteEconomia(idU: number, idEconomia: number): Observable<EsitoOperazioneEconomia> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/deleteeconomia';
        let params = new HttpParams().set("idU", idU.toString()).set("idEconomia", idEconomia.toString());
        return this.http.get<EsitoOperazioneEconomia>(url, { params: params });
    }

    modifica(idU: number, economiaDTO: EconomiaDTO): Observable<ModificaDTO> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/modifica';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.post<ModificaDTO>(url, economiaDTO, { params: params });
    }

    getBandi(idU: number): Observable<Array<BandoLinea>> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/getbandi';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.get<Array<BandoLinea>>(url, { params: params });
    }

    getProgettiByBando(idU: number, idBandoLinea: number): Observable<Array<ProgettoDTO>> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/getprogettibybando';
        let params = new HttpParams().set("idU", idU.toString()).set("idBandoLinea", idBandoLinea.toString());
        return this.http.get<Array<ProgettoDTO>>(url, { params: params });
    }

    cambiaProgetto(idU: number, idProgetto: number, idEconomia: number, idSoggettoBeneficiario: number, tipologiaProgetto: string): Observable<BeneficiarioQuoteDTO> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/cambiaprogetto';
        let params = new HttpParams()
            .set("idU", idU.toString())
            .set("idProgetto", idProgetto.toString())
            .set("tipologiaProgetto", tipologiaProgetto.toString());
        if (idEconomia) {
            params = params.set("idEconomia", idEconomia.toString());
        }
        if (idSoggettoBeneficiario) {
            params = params.set("idSoggettoBeneficiario", idSoggettoBeneficiario.toString());
        }
        return this.http.get<BeneficiarioQuoteDTO>(url, { params: params });
    }

    reportDettaglioEconomia(idU: number, request: ParametriRicercaEconomieDTO) {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/reportDettaglioEconomia';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.post(url, request, { params: params, responseType: 'blob', observe: 'response' });
    }

    salvaAggiornamentiEconomia(idU: number, request: ModificaDTO): Observable<EsitoOperazioneEconomia> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/salvaaggiornamentieconomia';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.post<EsitoOperazioneEconomia>(url, request, { params: params });
    }

    salvaNuovaEconomia(idU: number, request: ModificaDTO): Observable<EsitoOperazioneEconomia> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/salvanuovaeconomia';
        let params = new HttpParams().set("idU", idU.toString());
        return this.http.post<EsitoOperazioneEconomia>(url, request, { params: params });
    }

    infoProgetto(idU: number, idProgetto: number): Observable<DatiGeneraliDTO> {
        let url = this.configService.getApiURL() + '/restfacade/gestioneeconomie/infoProgetto';
        let params = new HttpParams().set("idU", idU.toString()).set("idProgetto", idProgetto.toString());
        return this.http.get<DatiGeneraliDTO>(url, { params: params });
    }
}

File: pbweb/pbweb-jar/src/main/java/it/csi/pbandi/pbweb/business/service/GestioneEconomieService.java
/*******************************************************************************
* Copyright Regione Piemonte - 2024
* SPDX-License-Identifier: EUPL-1.2
******************************************************************************/

package it.csi.pbandi.pbweb.business.service;

import javax.servlet.http.HttpServletRequest;
import javax.ws.rs.GET;
import javax.ws.rs.POST;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.Context;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import it.csi.pbandi.pbweb.dto.gestioneEconomica.ModificaDTO;
import it.csi.pbandi.pbweb.pbandisrv.dto.gestioneeconomie.EconomiaDTO;
import it.csi.pbandi.pbweb.pbandisrv.dto.gestioneeconomie.ParametriRicercaEconomieDTO;

@Path("/gestioneeconomie")
public interface GestioneEconomieService {

	@POST
	@Path("geteconomie")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getEconomie(@QueryParam("idU")Long idU, 
								ParametriRicercaEconomieDTO parametriRicercaEconomieDTO,
								@Context HttpServletRequest req) throws Exception;
	
	@GET
	@Path("deleteeconomia")
	@Produces(MediaType.APPLICATION_JSON)
	public Response deleteEconomia(@QueryParam("idU")Long idU, 
			                       @QueryParam("idEconomia") Long idEconomia,
							       @Context HttpServletRequest req) throws Exception;

	@POST
	@Path("modifica")
	@Produces(MediaType.APPLICATION_JSON)
	public Response modifica(@QueryParam("idU")Long idU, 
			                 EconomiaDTO economiaDTO,   //Passare solo idEconomia 
							 @Context HttpServletRequest req) throws Exception;
	
	@GET
	@Path("getbandi")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getBandi( @QueryParam("idU")Long idU, 
							  @Context HttpServletRequest req) throws Exception;
	
	
	//Viene invocato al cambio del bando aggiornando la lista dei progetti associati
	@GET
	@Path("getprogettibybando")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getProgettiByBando(@QueryParam("idU")Long idU, 
					                   @QueryParam("idBandoLinea") Long idBandoLinea,
									   @Context HttpServletRequest req) throws Exception;
	
	
	//Viene invocato al cambio del progetto aggiornando il beneficiario e le quote del progetto
	@GET
	@Path("cambiaprogetto")// DA MODIFICARE IN  cambiaprogetto
	@Produces(MediaType.APPLICATION_JSON)
	public Response cambiaProgetto(@QueryParam("idU")Long idU, 
									                 @QueryParam("idProgetto") Long idProgetto,
									                 @QueryParam("idEconomia") Long idEconomia,
									                 @QueryParam("idSoggettoBeneficiario") Long idSoggettoBeneficiario,
									                 @QueryParam("tipologiaProgetto") String tipologiaProgetto,  // tipologiaProgetto: C = cedente; R = ricevente.
													 @Context HttpServletRequest req) throws Exception;
	
	
	@GET
	@Path("infoProgetto")
	@Produces(MediaType.APPLICATION_JSON)
	public Response infoProgetto(@QueryParam("idU")Long idU, 
			                     @QueryParam("idProgetto") Long idProgetto,
							     @Context HttpServletRequest req) throws Exception;
	
	
//	@GET
//	@Path("gestionequote")
//	@Produces(MediaType.APPLICATION_JSON)
//	public Response gestioneQuote(@QueryParam("idU")Long idU, 
//								          @QueryParam("idProgetto") Long idProgetto,
//								          @QueryParam("idEconomia") Long idEconomia,
//								          @QueryParam("tipologiaProgetto") String tipologiaProgetto,  // tipologiaProgetto: C = cedente; R = ricevente.
//										  @Context HttpServletRequest req) throws Exception;
	
	
	@POST
	@Path("salvaaggiornamentieconomia")
	@Produces(MediaType.APPLICATION_JSON)
	public Response salvaAggiornamentiEconomia(@QueryParam("idU")Long idU, 
												ModificaDTO modificaDTO,
												@Context HttpServletRequest req) throws Exception;
	
	@POST
	@Path("salvanuovaeconomia")
	@Produces(MediaType.APPLICATION_JSON)
	public Response salvaNuovaEconomia(@QueryParam("idU")Long idU, 
												ModificaDTO modificaDTO,
												@Context HttpServletRequest req) throws Exception;
	
	
	@POST
	@Path("reportDettaglioEconomia") 
	@Produces(MediaType.APPLICATION_JSON)
	Response reportDettaglioEconomia(
									@QueryParam("idU")Long idU, 
									ParametriRicercaEconomieDTO parametriRicercaEconomieDTO,
									@Context HttpServletRequest req) throws Exception;
	
}
